File: game-theory/1561.py
# Google, Apple

# Three pointer technique
# Tc  = O(n.logn)
# Sc = O(1)

# Approach-1
class Solution:
    def maxCoins(self, piles: List[int]) -> int:
        n = len(piles)
        result = 0
        piles.sort()
        bob = 0
        myself = n - 2
        while myself > bob:
            result += piles[myself]
            myself -= 2
            bob += 1
        return result
    
class Solution:
    def maxCoins(self, piles: List[int]) -> int:
        n = len(piles)
        result = 0
        piles.sort()
        for M in range(n // 3, n, 2):
            result += piles[M]
        return result
    

File: game-theory/1561.cpp


class Solution {
public:
    int maxCoins(vector<int>& piles) {
        int n = piles.size();
        int result = 0;
        sort(begin(piles), end(piles));
        int bob = 0;
        int me = n - 2;
        // int alice = n - 1;
        while (me > bob) {
            result += piles[me];
            me -= 2;
            bob++;
            // alice -= 2;
        }
        return result;
    }
};

class Solution {
public:
    int maxCoins(vector<int>& piles) {
        int n = piles.size();
        int result = 0;
        sort(begin(piles), end(piles));
        for(int M = n/3; M < n; M += 2){
            result += piles[M];
        }
        return result;
    }
};